src/azul/service/avro_pfb.py

```diff
@@ -0,0 +1,451 @@
+from collections import (
+    defaultdict,
+)
+import logging
+from operator import (
+    attrgetter,
+)
+from typing import (
+    Iterable,
+    MutableMapping,
+    MutableSet,
+    cast,
+)
+import uuid
+
+import attr
+import fastavro
+from fastavro.validation import (
+    ValidationError,
+)
+from more_itertools import (
+    one,
+)
+
+from azul import (
+    config,
+)
+from azul.indexer.document import (
+    FieldTypes,
+    null_bool,
+    null_float,
+    null_int,
+    null_int_sum_sort,
+    null_str,
+    pass_thru_int,
+    pass_thru_json,
+)
+from azul.json_freeze import (
+    freeze,
+    sort_frozen,
+)
+from azul.plugins.metadata.hca.transform import (
+    pass_thru_uuid4,
+    value_and_unit,
+)
+from azul.types import (
+    JSON,
+    MutableJSON,
+)
+
+log = logging.getLogger(__name__)
+
+
+@attr.s(auto_attribs=True, frozen=True, kw_only=True)
+class Relation:
+    dst_id: str
+    # A more appropriate variable name would be dst_type, but we stick with name
+    # to conform to PFB spec
+    dst_name: str
+
+    @classmethod
+    def to_entity(cls, entity: 'PFBEntity'):
+        return cls(dst_id=entity.id, dst_name=entity.name)
+
+
+pfb_namespace = uuid.UUID('3372fc62-76d2-4ada-a68a-f08d4f558da5')
+
+
+@attr.s(auto_attribs=True, frozen=True, kw_only=True)
+class PFBEntity:
+    """
+    Python representation of the PFB data object. Attribute names conform to the
+    PFB spec (which simplifies serialization).
+    """
+    id: str
+    name: str
+    object: MutableJSON = attr.ib(eq=False)
+
+    @classmethod
+    def from_no_id(cls, name: str, object_: MutableJSON) -> 'PFBEntity':
+        """
+        Derive id from object in a reproducible way so that we can distinguish
+        entities by comparing their ids.
+
+        Entities need to be quickly distinguishable, but unfortunately
+        document_id wasn't unique (or present) for all entities. Thus, we use
+        the frozen state of the JSON to generate our hash and check equality.
+        """
+        entity_id = str(uuid.uuid5(pfb_namespace, repr(sort_frozen(freeze(object_)))))
+        return cls(id=entity_id, name=name, object=object_)
+
+    def to_json(self, relations: Iterable[Relation]):
+        return {
+            **attr.asdict(self),
+            "relations": [attr.asdict(relation) for relation in relations]
+        }
+
+
+class EntityAggregator:
```

Comment:
I don't think this class aggregates. Collate? Also, there already is a class of that exact name. So if we introduce a name collision, it should be for a good reason and the name should be spot on.

Reply:
After Googling "collate" I don't think that's accurate. No explicit ordering is done here.

Essentially this class transforms docs into normalize PFB entities. I'll try `PFBTransformer`